Msi 2030 add ability to mark orders placed with store pickup

swnsma opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
Add possibility to mark orders, created with 'In Store Pickup' Shipping method.
Add extension attribute to order entity with information about related pickup point.

### Fixed Issues
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento-engcom/msi#2030: Add an ability to Mark Orders placed with Store Pickup

### Points to discuss
Probably, it will be nice to add this extension attribute to collection join:
```
    <extension_attributes for="Magento\Sales\Api\Data\OrderInterface">
        <attribute code="pickup_point_id" type="string">
            <join reference_table="inventory_pickup_point_order" join_on_field="entity_id" reference_field="order_id">
                <field column="pickup_point_id">pickup_point_id</field>
            </join>
        </attribute>
```
But for this we need to add join processor for Order Repository 'getList' method in Magento 2 Core Repository (https://github.com/magento-engcom/msi/blob/475dc7c32e54621087b9669ca76ab1eacecde12f/app/code/Magento/Sales/Model/OrderRepository.php#L197)
E.g.
```
        /** @var \Magento\Framework\Api\ExtensionAttribute\JoinProcessor $extensionJoinProcessor */
        $extensionJoinProcessor = ObjectManager::getInstance()->get('\Magento\Framework\Api\ExtensionAttribute\JoinProcessor');
        $extensionJoinProcessor->process($searchResult, OrderInterface::class);
```
**UPD**: issue already reported https://github.com/magento/magento2/issues/8035

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
MSI-2030-Add an ability to Mark Orders placed with Store Pickup.
Declare db schema.

Commit:
MSI-2030-Add an ability to Mark Orders placed with Store Pickup.
Add new extension attribute.

Commit:
MSI-2030-Add an ability to Mark Orders placed with Store Pickup.
Decrease schema dependency on inventory_source.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/signed)](https://cla.dev.magento.com/magento-engcom/msi?pullRequest=2082) <br/>All committers have signed the CLA.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/ResourceModel/OrderPickupPoint/GetPickupPointByOrderId.php:
```diff
@@ -0,0 +1,60 @@
+use Magento\Framework\App\ResourceConnection;
+
+/**
+ * Class GetPickupPointByOrderId.
```
Repeating class name in the comment is redundant. Clear responsibility description is sufficient.

swnsma replied:
Done.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/ResourceModel/OrderPickupPoint/GetPickupPointByOrderId.php:
```diff
@@ -0,0 +1,60 @@
+ */
+class GetPickupPointByOrderId
+{
+    const ORDER_ID        = 'order_id';
```
Should these constants be `private`?

swnsma replied:
Done.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/ResourceModel/OrderPickupPoint/GetPickupPointByOrderId.php:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @return string|null
+     */
+    public function execute(int $orderId):?string
```
Why should pickup point id be a `string`? Maybe it's a point code?

swnsma replied:
Fixed.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/etc/db_schema.xml:
```diff
@@ -0,0 +1,18 @@
+-->
+<schema xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:framework:Setup/Declaration/Schema/etc/schema.xsd">
+    <table name="inventory_pickup_point_order" resource="default" engine="innodb">
```
We have to discuss `Pickup Point` naming. It's best to align it with business domain

swnsma replied:
Changed to 'Pickup Location'.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/etc/module.xml:
```diff
@@ -6,5 +6,10 @@
+    <module name="Magento_InventoryInStorePickup" setup_version="1.0.0">
+        <sequence>
+            <module name="Magento_Sales" />
+            <module name="Magento_Inventory" />
```
Should it be `Magento_Inventory` or `InventoryApi`?

Also `sales` should be added to composer dependencies

swnsma replied:
For now, dependency on Magento_Sales only.
Added to composer.json.

ishakhsuvarov reviewed (changes requested):
* `Pickup Point` term has to be discussed and aligned with business domain
* Please add integration tests as well
* Please update with the latest changes from `store-pickup` branch

Commit:
Merge remote-tracking branch 'origin/store-pickup' into MSI-2030-add-ability-to-mark-orders-placed-with-store-pickup

Commit:
MSI-2030-Add an ability to Mark Orders placed with Store Pickup.
Fix tests and apply code review corrections.

Commit:
Merge remote-tracking branch 'origin/store-pickup' into MSI-2030-add-ability-to-mark-orders-placed-with-store-pickup

Commit:
MSI-2030-Add an ability to Mark Orders placed with Store Pickup.
Add integration test for order placement.

MSI-2030-Add an ability to Mark Orders placed with Store Pickup.

Commit:
MSI-2030-Add an ability to Mark Orders placed with Store Pickup.
Pickup Point -> Pickup Location
Pickup Point Id -> Pickup Location Code

Commit:
MSI-2030-Add an ability to Mark Orders placed with Store Pickup.
Add join directive for extension attribute.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Test/_files/create_in_store_pickup_quote_on_eu_website.php:
```diff
@@ -0,0 +1,59 @@
+/** @var AddressInterface $address */
+$address = $addressFactory->create(
+    [
+        'data' => [
```
I love how you save a US address to the Quote in fixture designed for EU Website :)

ishakhsuvarov committed:
Merge branch 'store-pickup' into MSI-2030-add-ability-to-mark-orders-placed-with-store-pickup

ishakhsuvarov committed:
magento-engcom/msi#2082: Minor coding style fixes

ghost commented:
Hi @swnsma, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento-engcom+msi+2082), it will take less than a minute. </br>Your feedback will help us to improve contribution process.
